File: clients/typescript/tooling/generate.mjs
import { readFile, writeFile, mkdir } from "node:fs/promises";
import { fileURLToPath } from "node:url";
import { dirname, resolve } from "node:path";

const here = dirname(fileURLToPath(import.meta.url));
const packageRoot = resolve(here, "..");
const sourceSchema = resolve(
  packageRoot,
  "../../crates/runtrol-runtime-protocol/schema/runtime.schema.json",
);
const generatedTypes = resolve(packageRoot, "src/generated/protocol.ts");
const generatedSchema = resolve(packageRoot, "src/generated/schema.ts");
const packagedSchema = resolve(packageRoot, "schema/runtime.schema.json");
const checking = process.argv.includes("--check");

const schemaText = await readFile(sourceSchema, "utf8");
const schema = JSON.parse(schemaText);
const definitions = schema.$defs;
if (!definitions || typeof definitions !== "object" || Array.isArray(definitions)) {
  throw new Error("public Runtime schema has no object $defs");
}

function quoted(value) {
  return JSON.stringify(value);
}

function propertyName(name) {
  return /^[A-Za-z_$][A-Za-z0-9_$]*$/.test(name) ? name : quoted(name);
}

function typeOf(node) {
  if (node === true) return "unknown";
  if (node === false) return "never";
  if (!node || typeof node !== "object" || Array.isArray(node)) return "unknown";
  if (typeof node.$ref === "string") return node.$ref.split("/").at(-1);
  if (Object.hasOwn(node, "const")) return quoted(node.const);
  if (Array.isArray(node.enum)) return node.enum.map(quoted).join(" | ") || "never";
  if (Array.isArray(node.oneOf)) return node.oneOf.map(typeOf).join(" | ");
  if (Array.isArray(node.anyOf)) return node.anyOf.map(typeOf).join(" | ");
  if (Array.isArray(node.allOf)) return node.allOf.map(typeOf).join(" & ");
  if (Array.isArray(node.type)) {
    return node.type.map((one) => typeOf({ ...node, type: one })).join(" | ");
  }
  switch (node.type) {
    case "null":
      return "null";
    case "boolean":
      return "boolean";
    case "integer":
    case "number":
      return "number";
    case "string":
      return "string";
    case "array":
      return `ReadonlyArray<${typeOf(node.items ?? true)}>`;
    case "object":
      return objectType(node);
    default:
      return "unknown";
  }
}

function objectType(node) {
  const required = new Set(Array.isArray(node.required) ? node.required : []);
  const properties = node.properties && typeof node.properties === "object"
    ? Object.entries(node.properties)
    : [];
  const lines = properties.map(([name, value]) =>
    `readonly ${propertyName(name)}${required.has(name) ? "" : "?"}: ${typeOf(value)};`,
  );
  if (node.additionalProperties === true) {
    lines.push("readonly [key: string]: unknown;");
  } else if (node.additionalProperties && typeof node.additionalProperties === "object") {
    lines.push(`readonly [key: string]: ${typeOf(node.additionalProperties)};`);
  }
  if (lines.length === 0) return "Readonly<Record<string, never>>";
  return `{ ${lines.join(" ")} }`;
}

function documentation(node) {
  if (typeof node.description !== "string" || node.description.length === 0) return "";
  return `/** ${node.description.replaceAll("*/", "* /")} */\n`;
}

// The complete schema is a public documentation artifact. The SDK validator needs only the keywords it executes.
// Keeping those two representations separate avoids allocating descriptions, root catalogue aliases, release
// metadata, and other documentation on every Runtime client activation without weakening one wire check.
const validationKeywords = new Set([
  "$ref",
  "additionalProperties",
  "anyOf",
  "const",
  "enum",
  "format",
  "items",
  "maximum",
  "minimum",
  "oneOf",
  "properties",
  "required",
  "type",
]);

function compactValidationNode(node) {
  if (Array.isArray(node)) return node.map(compactValidationNode);
  if (!node || typeof node !== "object") return node;
  return Object.fromEntries(
    Object.entries(node)
      .filter(([key]) => validationKeywords.has(key))
      .filter(([key, value]) => {
        // The validator infers an object from its closed object keywords, and a constant already proves
        // its JSON type. Omitting those redundant type words keeps the complete checked projection inside
        // the existing activation memory budget as the public contract grows.
        if (key !== "type") return true;
        if (Object.hasOwn(node, "const")) return false;
        if (value !== "object") return true;
        return !Object.hasOwn(node, "properties")
          && !Object.hasOwn(node, "required")
          && !Object.hasOwn(node, "additionalProperties");
      })
      .map(([key, value]) => {
        if (key === "properties") {
          return [
            key,
            Object.fromEntries(
              Object.entries(value).map(([name, child]) => [name, compactValidationNode(child)]),
            ),
          ];
        }
        return [key, compactValidationNode(value)];
      }),
  );
}

function compactValidationSchema(sourceDefinitions) {
  return {
    $defs: Object.fromEntries(
      Object.entries(sourceDefinitions).map(([name, node]) => [name, compactValidationNode(node)]),
    ),
  };
}

const declarations = Object.entries(definitions).map(([name, node]) => {
  const docs = documentation(node);
  const plainObject = node && typeof node === "object" && !Array.isArray(node)
    && node.type === "object" && !node.oneOf && !node.anyOf && !node.allOf;
  if (!plainObject) return `${docs}export type ${name} = ${typeOf(node)};`;
  const body = objectType(node);
  if (body === "Readonly<Record<string, never>>") {
    return `${docs}export type ${name} = ${body};`;
  }
  return `${docs}export interface ${name} ${body}`;
});

const revisions = schema["x-runtrol-finalized-revisions"];
const limits = schema["x-runtrol-limits"];
if (!Array.isArray(revisions) || !limits || typeof limits !== "object") {
  throw new Error("public Runtime schema omits generated revision or limit inventory");
}

const generated = `// Generated from crates/runtrol-runtime-protocol/schema/runtime.schema.json. Do not edit.\n\n`
  + `export const FINALIZED_REVISIONS = ${JSON.stringify(revisions)} as const;\n`
  + `export const PUBLIC_LIMITS = ${JSON.stringify(limits, null, 2)} as const;\n\n`
  + `${declarations.join("\n\n")}\n`;
const validationSchema = compactValidationSchema(definitions);
const generatedSchemaText = "// Generated validation projection. The complete public schema remains in schema/runtime.schema.json. Do not edit.\n\n"
  + `export const VALIDATION_SCHEMA = ${JSON.stringify(validationSchema)} as const;\n`;

async function compare(path, expected, label) {
  let actual;
  try {
    actual = await readFile(path, "utf8");
  } catch {
    throw new Error(`${label} is missing; run npm run generate`);
  }
  if (actual !== expected) throw new Error(`${label} is stale; run npm run generate`);
}

if (checking) {
  await compare(generatedTypes, generated, "generated TypeScript protocol");
  await compare(generatedSchema, generatedSchemaText, "generated TypeScript schema document");
  await compare(packagedSchema, schemaText, "packaged Runtime schema");
} else {
  await mkdir(dirname(generatedTypes), { recursive: true });
  await mkdir(dirname(packagedSchema), { recursive: true });
  await writeFile(generatedTypes, generated, "utf8");
  await writeFile(generatedSchema, generatedSchemaText, "utf8");
  await writeFile(packagedSchema, schemaText, "utf8");
}

File: clients/python/tooling/generate.py
"""Generate Python type aliases and the packaged schema from the Rust-derived Runtime schema."""

from __future__ import annotations

import hashlib
import json
import keyword
import shutil
import sys
from pathlib import Path
from typing import Any

PACKAGE = Path(__file__).resolve().parents[1]
ROOT = PACKAGE.parents[1]
SOURCE = ROOT / "crates" / "runtrol-runtime-protocol" / "schema" / "runtime.schema.json"
SCHEMA = PACKAGE / "schema" / "runtime.schema.json"
GENERATED = PACKAGE / "python" / "runtrol_runtime" / "generated.py"


def reference(value: str) -> str:
    prefix = "#/$defs/"
    return f"ForwardRef({value[len(prefix):]!r})" if value.startswith(prefix) else "JsonValue"


def literal(value: object) -> str:
    return repr(value)


def expression(schema: dict[str, Any]) -> str:
    if "$ref" in schema:
        return reference(str(schema["$ref"]))
    if "const" in schema:
        return f"Literal[{literal(schema['const'])}]"
    if isinstance(schema.get("enum"), list):
        values = ", ".join(literal(value) for value in schema["enum"])
        return f"Literal[{values}]"
    variants = schema.get("oneOf") or schema.get("anyOf")
    if isinstance(variants, list):
        return " | ".join(expression(value) for value in variants if isinstance(value, dict)) or "JsonValue"
    kind = schema.get("type")
    if isinstance(kind, list):
        return " | ".join(expression({**schema, "type": value}) for value in kind)
    if kind == "string":
        return "str"
    if kind == "integer":
        return "int"
    if kind == "number":
        return "float"
    if kind == "boolean":
        return "bool"
    if kind == "null":
        return "None"
    if kind == "array":
        items = schema.get("items")
        return f"list[{expression(items) if isinstance(items, dict) else 'JsonValue'}]"
    if kind == "object":
        return "JsonObject"
    return "JsonValue"


def definition(name: str, schema: dict[str, Any]) -> str:
    properties = schema.get("properties")
    if schema.get("type") == "object" and isinstance(properties, dict):
        required = set(schema.get("required", []))
        fields = []
        for field, shape in properties.items():
            value = expression(shape) if isinstance(shape, dict) else "JsonValue"
            wrapper = "Required" if field in required else "NotRequired"
            fields.append(f"    {field!r}: {wrapper}[{value}],")
        body = "\n".join(fields)
        return f"{name} = TypedDict({name!r}, {{\n{body}\n}})"
    return f"{name}: TypeAlias = {expression(schema)}"


def render(schema: dict[str, Any], digest: str) -> str:
    definitions = schema.get("$defs", {})
    rendered = []
    for name, shape in sorted(definitions.items()):
        if not isinstance(name, str) or not name.isidentifier() or keyword.iskeyword(name):
            continue
        if isinstance(shape, dict):
            rendered.append(definition(name, shape))
    return "\n".join(
        [
            '"""Generated from the checked Rust Runtime schema. Do not edit by hand."""',
            "",
            "from __future__ import annotations",
            "",
            "from typing import ForwardRef, Literal, NotRequired, Required, TypeAlias, TypedDict",
            "",
            "JsonValue: TypeAlias = None | bool | int | float | str | list[\"JsonValue\"] | dict[str, \"JsonValue\"]",
            "JsonObject: TypeAlias = dict[str, JsonValue]",
            f"SCHEMA_SHA256 = {digest!r}",
            "",
            *rendered,
            "",
        ]
    )


def main() -> int:
    raw = SOURCE.read_bytes()
    schema = json.loads(raw)
    generated = render(schema, hashlib.sha256(raw).hexdigest())
    check = "--check" in sys.argv
    if check:
        if not SCHEMA.is_file() or SCHEMA.read_bytes() != raw:
            print("packaged Python schema is stale")
            return 2
        if not GENERATED.is_file() or GENERATED.read_text(encoding="utf-8") != generated:
            print("generated Python Runtime types are stale")
            return 2
        return 0
    SCHEMA.parent.mkdir(parents=True, exist_ok=True)
    GENERATED.parent.mkdir(parents=True, exist_ok=True)
    shutil.copyfile(SOURCE, SCHEMA)
    GENERATED.write_text(generated, encoding="utf-8", newline="\n")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
